mypyc/lib-rt/attr.c

```diff
@@ -0,0 +1,173 @@
+// Generic native class attribute getters and setters
+
+#include <Python.h>
+#include "CPy.h"
+
+PyObject *CPyAttr_UndefinedError(PyObject *self, CPyAttr_Context *context) {
+    assert(!context->always_defined && "attribute should be initialized!");
+    PyErr_Format(PyExc_AttributeError,
+        "attribute '%s' of '%s' undefined", context->attr_name, Py_TYPE(self)->tp_name);
+    return NULL;
+}
+
+int CPyAttr_UndeletableError(PyObject *self, CPyAttr_Context *context) {
+    PyErr_Format(PyExc_AttributeError,
+        "'%s' object attribute '%s' cannot be deleted", Py_TYPE(self)->tp_name, context->attr_name);
+    return -1;
+}
+
+static void set_bitmap(PyObject *self, CPyAttr_Context *context, bool defined) {
+    uint32_t *bitmap = (uint32_t *)((char *)self + context->bitmap.offset);
+    if (defined) {
+        *bitmap |= context->bitmap.mask;
+    } else {
+        *bitmap &= ~context->bitmap.mask;
+    }
+}
+
+static inline bool is_undefined_via_bitmap(PyObject *self, CPyAttr_Context *context) {
+    return !(*(uint32_t *)((char *)self + context->bitmap.offset) & context->bitmap.mask);
+}
+
+PyObject *CPyAttr_GetterPyObject(PyObject *self, CPyAttr_Context *context) {
+    PyObject *value = *(PyObject **)((char *)self + context->offset);
+    if (unlikely(value == NULL)) {
+        return CPyAttr_UndefinedError(self, context);
+    }
+    return Py_NewRef(value);
+}
+
+PyObject *CPyAttr_GetterTagged(PyObject *self, CPyAttr_Context *context) {
+    CPyTagged value = *(CPyTagged *)((char *)self + context->offset);
+    if (unlikely(value == CPY_INT_TAG)) {
+        return CPyAttr_UndefinedError(self, context);
+    }
+    return CPyTagged_AsObject(value);
+}
+
+PyObject *CPyAttr_GetterBool(PyObject *self, CPyAttr_Context *context) {
+    char value = *((char *)self + context->offset);
+    if (unlikely(value == 2)) {
+        return CPyAttr_UndefinedError(self, context);
+    }
+    return Py_NewRef(value ? Py_True : Py_False);
+}
+
+PyObject *CPyAttr_GetterFloat(PyObject *self, CPyAttr_Context *context) {
+    double value = *(double *)((char *)self + context->offset);
+    if (value == CPY_FLOAT_ERROR
+            && !context->always_defined
+            && is_undefined_via_bitmap(self, context)) {
+        return CPyAttr_UndefinedError(self, context);
+    }
+    return PyFloat_FromDouble(value);
+}
+
+PyObject *CPyAttr_GetterInt32(PyObject *self, CPyAttr_Context *context) {
```

Comment:
There's no good reason for why there are no generic fixed-width integer setters other than I got lazy (:slightly_smiling_face:). Let me know if you want me to add them (or remove these getters alternatively).

Reply:
They've been removed: https://github.com/python/mypy/pull/15336/commits/e6d83cecd066cb1b513673344d6ccfbcf24654fa.